test/unittest/reduction/reduction_test.cpp

```diff
@@ -0,0 +1,181 @@
+/***************************************************************************
+ *
+ *  @license
+ *  Copyright (C) Codeplay Software Limited
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  For your convenience, a copy of the License has been included in this
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename reduction_test.cpp
+ *
+ **************************************************************************/
+
+#include <limits>
+
+#include "blas_test.hpp"
+#include <operations/blas_operators.h>
+#include <iostream>
+
+enum operator_t : int {
+  Add = 0,
+  Product = 1,
+  Division = 2,
+  Max = 3,
+  Min = 4,
+  AbsoluteAdd = 5
+};
+
+using index_t = int;
+
+template <typename scalar_t>
+using combination_t = std::tuple<index_t, index_t, index_t, operator_t>;
+
+/* Note: the product and division are not tested because our random data may
+ * contain values close to zero */
+const auto combi = ::testing::Combine(
+    ::testing::Values(1, 7, 513),                // rows
+    ::testing::Values(1, 15, 1000, 1337, 8195),  // columns
+    ::testing::Values(1, 2, 3),                  // ld_mul
+    ::testing::Values(operator_t::Add, operator_t::Max, operator_t::Min,
+                      operator_t::AbsoluteAdd));
+
+template <typename scalar_t>
+void run_test(const combination_t<scalar_t> combi) {
+  index_t rows, cols, ld_mul;
+  operator_t op;
+  std::tie(rows, cols, ld_mul, op) = combi;
+
+  using data_t = utils::data_storage_t<scalar_t>;
+
+  auto q = make_queue();
+  test_executor_t ex(q);
+
+  auto policy_handler = ex.get_policy_handler();
+
+  index_t ld = rows * ld_mul;
+
+  std::vector<data_t> in_m(ld * cols);
+  std::vector<data_t> out_v_gpu(rows);
+  std::vector<data_t> out_v_cpu(rows);
+
+
+
+  fill_random(in_m);
+  for (index_t i = 0; i < rows; i++) {
+    out_v_gpu[i] = -1;
+  }
+  scalar_t init_val;
+  switch (op) {
+    case operator_t::Add:
+    case operator_t::AbsoluteAdd:
+      init_val = 0;
+      break;
+    case operator_t::Product:
+    case operator_t::Division:
+      init_val = 1;
+      break;
+    case operator_t::Min:
+      init_val = std::numeric_limits<scalar_t>::max();
+      break;
+    case operator_t::Max:
+      init_val = std::numeric_limits<scalar_t>::min();
```

Comment:
I believe the identity for max operations is `lowest()` rather than `min()`. Using `min()` I think causes incorrect behaviour with floating point types, for integers it's fine as `min() == lowest()`.

Reply:
Changing to lowest causes the max reduction tests to fail. Keeping as `min()` for now

Reply:
Surely this indicates a problem in the max reduction tests to be fixed then?

Reply:
```qoute 
lowest() is necessary because not all floating-point representations have a smallest (most negative) value that is the negative of the largest (most positive) finite value.
```
If it does not work with `lowest()`, it means that there is a bug in the test that is needed to be fixed.

Reply:
Fixed the issue. The tests work with `lowest()` now.